dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -348,6 +334,30 @@ class LocalTunnelSenderV2 : public TunnelSender
 private:
     friend class tests::TestMPPTunnel;
 
+    template <bool non_blocking>
+    bool pushImpl(TrackedMppDataPacketPtr && data)
+    {
+        if (unlikely(checkPacketErr(data)))
+            return false;
+
+        // receiver_mem_tracker pointer will always be valid because ExchangeReceiverBase won't be destructed
+        // before all local tunnels are destructed so that the MPPTask which contains ExchangeReceiverBase and
+        // is responsible for deleting receiver_mem_tracker must be destroyed after these local tunnels.
+        data->switchMemTracker(local_request_handler.recv_mem_tracker);
+
+        // When ExchangeReceiver receives data from local and remote tiflash, number of local tunnel threads
+        // is very large and causes the time of transfering data by grpc threads becomes longer, because
+        // grpc thread is hard to get chance to push data into MPMCQueue in ExchangeReceiver.
+        // So, we need to control the concurrency of local tunnel threads by the lock.
```

Comment:
Better to add more comments.
```suggestion
        // Adding a lock ensures that there is only one other thread competing with async reactor, so the probability of async reactor getting the lock is 1/2.
        // If the number of local tunnel sender threads is n, then the probability of a local tunnel sender thread getting the lock 
decreases from 1/(1+n) to 1/2 * 1/n = 1/2n
```

Reply:
> Better to add more comments.

okk